dask/array/tests/test_routines.py

```diff
@@ -1549,22 +1549,49 @@ def test_coarsen_with_excess():
     )
 
 
-def test_coarsen_bad_chunks():
-
-    x1 = da.arange(10, chunks=5)
-    x2 = x1.rechunk((1, 2, 3, 4))
-    assert_eq(da.coarsen(np.sum, x1, {0: 5}), da.coarsen(np.sum, x2, {0: 5}))
+@pytest.mark.parametrize("chunks", [(x,) * 3 for x in range(16, 32)])
+def test_coarsen_bad_chunks(chunks):
+    x1 = da.arange(np.sum(chunks), chunks=5)
+    x2 = x1.rechunk(tuple(chunks))
+    assert_eq(
+        da.coarsen(np.sum, x1, {0: 10}, trim_excess=True),
+        da.coarsen(np.sum, x2, {0: 10}, trim_excess=True),
+    )
 
 
-def test_aligned_coarsen_chunks():
+@pytest.mark.parametrize(
+    "chunks, divisor",
+    [
+        ((1, 1), 1),
+        ((1, 1), 2),
+        ((1, 1, 1), 2),
+        ((10, 1), 10),
+        ((20, 10, 15, 23, 24), 10),
+        ((20, 10, 15, 23, 24), 8),
+        ((10, 20, 30, 40, 2), 10),
+        ((20, 10, 15, 42, 23, 24), 16),
+        ((20, 10, 15, 47, 23, 24), 10),
+        ((2, 10, 15, 47, 23, 24), 4),
+    ],
+)
+def test_aligned_coarsen_chunks(chunks, divisor):
 
     from ..routines import aligned_coarsen_chunks as acc
 
-    assert acc((20, 10, 15, 23, 24), 10) == (20, 10, 20, 20, 20, 2)
-    assert acc((20, 10, 15, 42, 23, 24), 10) == (20, 10, 20, 40, 20, 20, 4)
-    assert acc((20, 10, 15, 47, 23, 24), 10) == (20, 10, 20, 50, 20, 10, 9)
-    assert acc((2, 10, 15, 47, 23, 24), 10) == (10, 20, 50, 20, 20, 1)
-    assert acc((10, 20, 30, 40, 2), 10) == (10, 20, 30, 40, 2)
+    aligned_chunks = acc(chunks, divisor)
+    any_remainders = (np.array(aligned_chunks) % divisor) != 0
+
+    # check that total number of elements is conserved
+    assert sum(aligned_chunks) == sum(chunks)
+    # check that no chunks are 0
+    assert (np.array(aligned_chunks) > 0).all()
+    # check that at most one chunk was added
+    assert len(aligned_chunks) <= len(chunks) + 1
+    # check that either 0 or 1 chunks are not divisible by divisor
+    assert any_remainders.sum() in (0, 1)
+    # check that the only indivisible chunk is the last
+    if any_remainders.sum() == 1:
+        assert any_remainders[-1] == 1
```

Comment:
Instead of checking that `aligned_coarsen_chunks` produces an exact sequence of chunks, I check for whether the chunks it produces satisfy a set of criteria, because I don't think the functionality of `aligned_coarsen_chunks` should be pinned to a specific chunk order. 

E.g., my implementation of`aligned_coarsen_chunks(20, 10, 15, 23, 24), 10)` returns `(20, 20, 10, 20, 20, 2)` but a different implementation could resonably return`(20, 10, 20, 20, 20, 2)`, so this test doesn't distinguish between these two outputs.

Reply:
> (20, 20, 10, 20, 20, 2) but a different implementation could resonably return (20, 10, 20, 20, 20, 2)

Note that the first one touches an extra block though (the second one) which means more communication. Also a simpler fix would be to replace the hardcoded `5` by `multiple // 2`. That fixes your original issue.

Reply:
I tried the `multiple // 2` change and new tests still failed, which is why I did a re-write of the original. If the original code can be tweaked to pass all the new tests, then I'm more than happy to ditch my rewrite.

Reply:
and I can also adjust the tests to properly check if extra blocks get touched